node/network/bitfield-distribution/src/lib.rs

```diff
@@ -0,0 +1,684 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The bitfield distribution
+//!
+//! In case this node is a validator, gossips its own signed availability bitfield
+//! for a particular relay parent.
+//! Independently of that, gossips on received messages from peers to other interested peers.
+
+use codec::{Decode, Encode};
+use futures::{channel::oneshot, FutureExt};
+
+use node_primitives::{ProtocolId, View};
+
+use log::{debug, info, trace, warn};
+use polkadot_node_subsystem::messages::*;
+use polkadot_node_subsystem::{
+	FromOverseer, OverseerSignal, SpawnedSubsystem, Subsystem, SubsystemContext, SubsystemResult,
+};
+use polkadot_primitives::v0::{SigningContext, ValidatorId};
+use polkadot_primitives::v1::{Hash, SignedAvailabilityBitfield};
+use sc_network::ReputationChange;
+use std::collections::{HashMap, HashSet};
+
+const COST_SIGNATURE_INVALID: ReputationChange =
+	ReputationChange::new(-100, "Bitfield signature invalid");
+const COST_VALIDATOR_INDEX_INVALID: ReputationChange =
+	ReputationChange::new(-100, "Bitfield validator index invalid");
+const COST_MISSING_PEER_SESSION_KEY: ReputationChange =
+	ReputationChange::new(-133, "Missing peer session key");
+const COST_NOT_INTERESTED: ReputationChange =
+	ReputationChange::new(-51, "Not intersted in that parent hash");
+const COST_MESSAGE_NOT_DECODABLE: ReputationChange =
+	ReputationChange::new(-100, "Not intersted in that parent hash");
+
+/// Checked signed availability bitfield that is distributed
+/// to other peers.
+#[derive(Encode, Decode, Debug, Clone, PartialEq, Eq)]
+pub struct BitfieldGossipMessage {
+	/// The relay parent this message is relative to.
+	pub relay_parent: Hash,
+	/// The actual signed availability bitfield.
+	pub signed_availability: SignedAvailabilityBitfield,
+}
+
+/// Data used to track information of peers and relay parents the
+/// overseer ordered us to work on.
+#[derive(Default, Clone)]
+struct Tracker {
+	/// track all active peers and their views
+	/// to determine what is relevant to them.
+	peer_views: HashMap<PeerId, View>,
+
+	/// Our current view.
+	view: View,
+
+	/// Additional data particular to a relay parent.
+	per_relay_parent: HashMap<Hash, PerRelayParentData>,
+}
+
+/// Data for a particular relay parent.
+#[derive(Debug, Clone, Default)]
+struct PerRelayParentData {
+	/// Signing context for a particular relay parent.
+	signing_context: SigningContext<Hash>,
```

Comment:
```suggestion
	signing_context: SigningContext,
```

The generic is just for the benefit of the runtime